### PR TITLE
Add edit food item screen

### PR DESCRIPTION
**Pull request by matvitn**
**Description:**

This pull request encompasses several significant improvements and updates to the codebase, focusing on adding an edit screen for food items, refactoring, increasing functionality, and bolstering test coverage. Read below for a detailed summary of all the changes:

---

### **1. Added EditFoodItemScreen**

- **Created New Screen that Allows Users to Modify Previously Added Food Items:**
  - Users can now modify the amount, location and dates for food items that are in their household

- **Similar To AddFoodScreen:**
  - The Logic and UI for this new screen is very similar to AddFoodItem (except it starts with previous inputs and only allows certain fields to be modified) ensuring consistency across the app and functionality (since AddFoodItem works properly)

- **Added New Function In HouseHoldViewModel to Allow Modification to FoodItems:**
  - Similar to addFoodItem, a new function called editFoodItem has been added to HouseHoldViewModel which removes the old foodItem and adds the new one to the repository

- **Affected Files:**
  - `EditFoodItemScreen.kt`
  - `HouseHoldViewModel.kt` (adjusted for compatibility)

---

### **2. Refactored Utility Functions**

- **Moved Input validation Utility Functions to Utils Folder:**
  - Extracted utility functions from `AddFoodItem` and relocated them to the `utils` package.
  - This promotes code reusability and better organization, allowing multiple screens to share common functionality.

- **Affected Files:**
  - `AddFoodItem.kt` (utility functions removed)
  - `InputValidationUtils.kt` (new file in `utils` containing input fields validation utility functions)

---

### **3. Added Tests for EditFoodItemScreen**

- **Developed UI Tests Using Compose Testing Framework:**
  - Created `EditFoodItemScreenTest.kt` to cover the user interactions and validations within the `EditFoodItemScreen`. These tests are very similar to AddFoodItemScreenTest since the screens behave similarly
  - Tests include:
    - Initial UI components display.
    - Input validation for amount field.
    - Dropdown selections for location.
    - Date field validations.
    - Form submission with both valid and invalid data.
    - Navigation actions (e.g., back button and cancel button functionality).

- **Adjusted Tests for Reliability:**
  - Updated tests to use specific matchers and reliable identifiers (e.g., `testTag`) to improve stability.

- **Affected Files:**
  - `EditFoodItemScreenTest.kt` (new file containing UI tests)

---

### **4. Modified IndividualFoodItem**

- **Fixed Conceptual issue:**
  - Changed parameter from HouseHoldViewModel to ListFoodItemViewModel since all selecting food item logic should be done from there.
  - Removed function and variable performing food item selection from HouseHoldView Model and ensured any logic that needs to select a food item uses ListFoodItemsViewModel instead

- **Affected Files:**
  - `IndividualFoodItem.kt`
  - `HouseholdViewModel.kt`

---

### **5. Updated IndividualFoodItem and HouseholdViewModel tests**

- **Aligned IndividualFoodItem Tests with New Functionality:**
  - Modified `IndividualFoodItemTests` to reflect changes in the `IndividualFoodItem`.
  - Ensured that the HouseholdViewModel tests cover the new editFoodItem function.
  - Confirmed that all tests pass, providing confidence in the application's behavior.

- **Affected Files:**
  - `IndividualFoodItemTest.kt`
  - `HouseholdViewModelTest.kt`

---

### **Additional Improvements**

- **Added Function for Modifying existing foodItem in HouseHoldViewModel:**
  - editFoodItem (similarly to addFoodItem) removes oldFoodItem (passed as parameter) from the repository and adds newFoodItem (passed as parameter)

- **Formatted Code:**
  - Applied ktfmtFormat to the project to ensure all code is properly formatted.

---

### **Testing and Verification**

- **Unit Tests:**
  - All unit tests pass, confirming the correctness of utility functions and validation logic.

- **UI Tests:**
  - All UI tests pass, verifying that the screens behave as expected under various user interactions.

- **End-to-End Tests:**
  - End-to-end tests pass, demonstrating the application's stability and reliability.

---

### **Not Yet Implemented**
- **No Navigation to EditFoodItemScreen from App Yet:**
  - As it stands, despite it being properly implemented and tested, it is not yet possible to naturally navigate to EditFoodItemScreen through the app as I this PR is already quite large. Navigation is in development right now and a new PR implementing the feature should be ready for review soon

---

### **Conclusion**

This pull request significantly enhances the codebase by:

- Adding a new screen that provides new functionality to the user
- Promoting code reuse and maintainability through refactoring.
- Ensuring consistent and robust input validation across screens.
- Strengthening test coverage to catch issues early and improve reliability.

---

**Please review the changes and provide feedback.**

---

**Notes for Reviewers:**

- Focus on the integration of the new utility functions and the consistency of the new screen.
- Verify that the tests cover all critical paths and handle edge cases effectively.
- Ensure that the UI components are testable, with proper semantics and test tags.

---

Thank you for your time and consideration in reviewing this pull request.

### CODE REVIEW

File: app/src/main/java/com/android/shelfLife/ui/overview/EditFoodItem.kt [hunk 1]
```diff
@@ -0,0 +1,327 @@
...
+                              textAlign = TextAlign.Start)
+                        }
+                      }
+                      Spacer(modifier = Modifier.width(8.dp))
+                      // Unit
+                      // TODO this doesnt look good on the UI
+                      Card(
+                          border = CardDefaults.outlinedCardBorder(),
+                          modifier = Modifier.testTag("editFoodUnit")) {
+                            Text(text = selectedFood.foodFacts.quantity.unit.name)
```
Comment (alextheprophecy) [inline app/src/main/java/com/android/shelfLife/ui/overview/EditFoodItem.kt:155]: Fix this Todo with better padding around card and a vertical alignement to center it.

```suggestion
                    Card(
                        border = CardDefaults.outlinedCardBorder(),
                        modifier = Modifier
                          .testTag("editFoodUnit")
                          .padding(horizontal = 8.dp, vertical = 4.dp).align(Alignment.CenterVertically)
                      ) {
                        Text(
                          text = selectedFood.foodFacts.quantity.unit.name,
                          modifier = Modifier.padding(8.dp)
                        )
                      }
```

Comment (alextheprophecy) [reply app/src/main/java/com/android/shelfLife/ui/overview/EditFoodItem.kt:155]: ### UX NOTE
- In the case of a user error (typing ML instead of GRAMS) on a foodItem creation, this could hinder UX since the user can't fix their mistake in the edit screen (unmodifiable card value)
- On the other hand, stopping the user from changing the Unit is good, since for example a Liter of milk should never need to be measured in grams in the future.

- **To be discussed to find a better solution for UX**

File: app/src/main/java/com/android/shelfLife/ui/overview/EditFoodItem.kt [hunk 2]
```diff
@@ -0,0 +1,327 @@
...
+                          }
+                    }
+                Spacer(modifier = Modifier.height(16.dp))
+              }
+
+              item(key = "location") {
+                // Location Dropdown
+                DropdownFields(
```
Comment (alextheprophecy) [inline app/src/main/java/com/android/shelfLife/ui/overview/EditFoodItem.kt:163]: Good use of our util dropdown composable

File: app/src/main/java/com/android/shelfLife/ui/overview/EditFoodItem.kt [hunk 3]
```diff
@@ -0,0 +1,327 @@
...
+                      } else {
+                        Toast.makeText(
+                                context,
+                                "Please correct the errors before submitting.",
+                                Toast.LENGTH_SHORT)
+                            .show()
+                      }
+                    },
+                    modifier = Modifier.testTag("foodSave").fillMaxWidth().height(50.dp)) {
+                      Text(text = "Submit", fontSize = 18.sp)
```
Comment (alextheprophecy) [inline app/src/main/java/com/android/shelfLife/ui/overview/EditFoodItem.kt:309]: ### Hardcoded strings
We use the strings resource in `res/values/strings.xml`. Consider removing the hardcoded strings here for better localisation (and potential translations).

- "Submit"
- "Cancel"

File: app/src/main/java/com/android/shelfLife/ui/overview/EditFoodItem.kt [hunk 4]
```diff
@@ -0,0 +1,327 @@
...
+  }
+
+  Scaffold(
+      modifier = Modifier.fillMaxSize(),
+      topBar = {
+        androidx.compose.material3.TopAppBar(
+            title = {
+              Text(
+                  modifier = Modifier.testTag("editFoodItemTitle"),
+                  text = stringResource(id = R.string.add_food_item_title))
```
Comment (alextheprophecy) [inline app/src/main/java/com/android/shelfLife/ui/overview/EditFoodItem.kt:106]: Should modify the string resource to point to a "Edit Food item" string instead of the Add food item (not yet implemented in the string resource)
```suggestion
                  text = stringResource(id = R.string.edit_food_item_title))
```

File: app/src/main/java/com/android/shelfLife/ui/overview/EditFoodItem.kt [hunk 5]
```diff
@@ -0,0 +1,327 @@
...
+                    .padding(innerPadding)
+                    .testTag("editFoodItemScreen"),
+            horizontalAlignment = Alignment.CenterHorizontally,
+            verticalArrangement = Arrangement.Top) {
+              item(key = "amount") {
+                Row(
+                    modifier = Modifier.fillMaxWidth(),
+                    horizontalArrangement = Arrangement.SpaceBetween) {
+                      Column(modifier = Modifier.weight(1f)) {
+                        // Amount Field with Error Handling
```
Comment (alextheprophecy) [inline app/src/main/java/com/android/shelfLife/ui/overview/EditFoodItem.kt:129]: ### Improve UX
showing the food item name at the top of the page would greatly improve UX, to reassure the user they pressed the right item, and to remember what they are modifying exactly
```suggestion
                        // Amount Field with Error Handling
                        Text(
                            text = selectedFood.foodFacts.name,
                            style = MaterialTheme.typography.headlineMedium,
                            modifier = Modifier.testTag("editFoodName").align(Alignment.Start).padding(8.dp))
```

File: app/src/main/java/com/android/shelfLife/ui/utils/InputValidationUtils.kt
Comment (alextheprophecy) [inline app/src/main/java/com/android/shelfLife/ui/utils/InputValidationUtils.kt:1]: Good job on moving these validation functions to a seperate file. 

> **FUTURE TASK:** We need to start using string resources more, since all the error messages in this file are not using localisation

**Comment by alextheprophecy** [comment]
## **Good Job!**  
> **Note**: there may be a lot of added lines, but many of them are extremely similar to AddFoodItem, and the corresponding test file. 

- **Overall excellent work** on maintaining high code quality and testing. 

- **View Model Updates**:  You noticed some flaws and deprecated ViewModel usage which you correctly updated and made sure to update the tests accordingly.

- **Code Coverage**: SonarCloud reports 99.5% line coverage on `EditFoodItemScreen` — impressive stakhanovite behaviour 😂!

- **Validation Separation**: Nice work on extracting input validation to a separate file — this is a clean, modular approach that enhances readability and maintainability.

- **Nice tweaks in codebase** (e.g ListFoodItemsViewModel: Allowing the unselection of the foodItem stateFlow in ListFoodItemsViewModel)

---

## **Future Considerations**

- **Duplication in Edit/Add Screens**: As we have discussed together, there is substantial similarity between `EditFoodItem` and `AddFoodItem`, which has prompted duplication warnings. Modularizing the skeleton of these screens, as you mentioned, is an excellent plan to reduce redundancy and streamline the codebase.

- **UI/UX modifications**: I have left some comments for UX and UI improvements in the EditFoodItem. These modifications are important, but can wait, since the Modularisation of Add/Edit FoodItem has not been done yet. -> Make sure to apply these changes during that task!

- **String Resource**: in general, our codebase contains many hard-coded strings. A future task could handle the string resources and make sure to allow proper localisation

---


## How the UX was tested on the PR
- **Manual Testing**: (You correctly identified that the screen navigation was not ready yet) So I performed manual testing by replacing `IndividualFoodItemScreen` with `EditFoodItemScreen` in the `MainActivity` under the `INDIVIDUAL_FOOD_ITEM` navigation route. This approach enabled me to update the selected food item from the Overview screen and validate the edit screen functionality directly on my mobile device.

---

### Final Note
Overall, this is a solid PR with well-considered changes and very high test coverage.  We have discussed the future plans for the duplications. Great work!

**Review by alextheprophecy — APPROVED** [review]
